physionet-django/project/models.py

```diff
@@ -1533,28 +1533,179 @@ def get_storage_info(self, force_calculate=True):
             compressed_used=compressed)
 
     def citation_text(self):
+        """
+        Returns the citation of the project in the following styles.
+
+        Parameters
+        ----------
+        styles (list):
+            MLA : MLA (8th edition)
+            APA (default) : APA (7th edition)
+            Chicago : Chicago (17th edition)
+            Harvard : Harvard
+            Vancouver : Vancouver
+
+        Returns
+        -------
+        citation_dict (dictionary) : dictionary of each citation style and
+            their respective citation for the requested project
+        """
+        # For if there are no authors or valuable metadata
         if self.is_legacy:
-            return ''
+            return {'APA':''}
 
+        styles = ['APA','MLA','Chicago','Harvard','Vancouver']
         authors = self.authors.all().order_by('display_order')
-        if self.doi:
-            return format_html(
-                '{authors} ({year}). {title} (version {version}). '
-                '<i>PhysioNet</i>. '
-                '<a href="https://doi.org/{doi}">https://doi.org/{doi}</a>',
-                authors=', '.join(a.initialed_name() for a in authors),
-                year=self.publish_datetime.year,
-                title=self.title,
-                version=self.version,
-                doi=self.doi)
-        else:
-            return format_html(
-                '{authors} ({year}). {title} (version {version}). '
-                '<i>PhysioNet</i>.',
-                authors=', '.join(a.initialed_name() for a in authors),
-                year=self.publish_datetime.year,
-                title=self.title,
-                version=self.version)
+        citation_dict = {}
+
+        for style in styles:
+
+            if (style == 'MLA'):
+
+                if (len(authors) == 1):
+                    temp_authors = authors[0].get_full_name(reverse=True)
+                elif (len(authors) == 2):
+                    first_author = authors[0].get_full_name(reverse=True)
+                    second_author = authors[1].get_full_name()
+                    temp_authors = first_author + ', and ' + second_author
+                else:
+                    temp_authors = ', '.join(authors[0].get_full_name().split())
+                    temp_authors += ', et al'
+
+                if self.doi:
```

Comment:
the only difference here appears to be presence/absence of the doi. Is there a way to avoid the code duplication (and ideally the if clause too?).

Reply:
I noticed that as well and have been thinking about it for a while but I'll try to focus on it to see if I can come up with a solution... this issue is the formatting string would also have to change.

Reply:
Maybe append the DOI part after the initial template is created?